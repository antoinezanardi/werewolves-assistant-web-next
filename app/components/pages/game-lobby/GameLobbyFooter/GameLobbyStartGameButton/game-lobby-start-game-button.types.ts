type GameLobbyStartGameButtonEmits = {
  rejectPlayersPositionStep: [];
};

export type {
  GameLobbyStartGameButtonEmits,
};