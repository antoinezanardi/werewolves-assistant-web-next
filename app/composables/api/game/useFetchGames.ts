import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchGames = {
  createGame: (createGameDto: CreateGameDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  getGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  cancelGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  makeGamePlay: (gameId: string, makeGamePlayDto: MakeGamePlayDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  createGameFeedback: (gameId: string, createGameFeedbackDto: CreateGameFeedbackDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
};

function useFetchGames(): UseFetchGames {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function createGame(createGameDto: CreateGameDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games`, {
        method: "POST",
        body: JSON.stringify(createGameDto),
      });
    } catch {
      return null;
    }
  }

  async function getGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "GET" });
    } catch {
      return null;
    }
  }

  async function cancelGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "DELETE" });
    } catch {
      return null;
    }
  }

  async function makeGamePlay(gameId: string, makeGamePlayDto: MakeGamePlayDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/play`, {
        method: "POST",
        body: JSON.stringify(makeGamePlayDto),
      });
    } catch {
      return null;
    }
  }

  async function createGameFeedback(gameId: string, createGameFeedbackDto: CreateGameFeedbackDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/feedback`, {
        method: "POST",
        body: JSON.stringify(createGameFeedbackDto),
      });
    } catch {
      return null;
    }
  }
  return {
    createGame,
    getGame,
    cancelGame,
    makeGamePlay,
    createGameFeedback,
  };
}

export { useFetchGames };