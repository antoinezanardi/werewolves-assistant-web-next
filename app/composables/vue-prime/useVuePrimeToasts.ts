import type { ToastMessageOptions } from "primevue/toast";
import { useToast } from "primevue/usetoast";

import { DEFAULT_VUE_PRIME_TOAST_OPTIONS } from "~/composables/vue-prime/constants/vue-prime.constants";

type UseVuePrimeToasts = {
  addToast: (options: ToastMessageOptions) => void;
  addSuccessToast: (options: ToastMessageOptions) => void;
  addInfoToast: (options: ToastMessageOptions) => void;
  addWarnToast: (options: ToastMessageOptions) => void;
  addErrorToast: (options: ToastMessageOptions) => void;
};

function useVuePrimeToasts(): UseVuePrimeToasts {
  const toast = useToast();

  function addToast(options: ToastMessageOptions): void {
    toast.add({
      ...DEFAULT_VUE_PRIME_TOAST_OPTIONS,
      ...options,
    });
  }

  function addSuccessToast(options: ToastMessageOptions): void {
    addToast({ severity: "success", ...options });
  }

  function addInfoToast(options: ToastMessageOptions): void {
    addToast({ severity: "info", ...options });
  }

  function addWarnToast(options: ToastMessageOptions): void {
    addToast({ severity: "warn", ...options });
  }

  function addErrorToast(options: ToastMessageOptions): void {
    addToast({ severity: "error", ...options });
  }
  return {
    addToast,
    addSuccessToast,
    addInfoToast,
    addWarnToast,
    addErrorToast,
  };
}

export { useVuePrimeToasts };