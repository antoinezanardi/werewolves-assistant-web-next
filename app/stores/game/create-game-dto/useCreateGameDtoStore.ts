import { defineStore } from "pinia";
import { ref } from "vue";
import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import type { CreateGameAdditionalCardDto } from "~/composables/api/game/dto/create-game/create-game-additional-card/create-game-additional-card.dto";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { GameAdditionalCardRecipientRoleName } from "~/composables/api/game/types/game-additional-card/game-additional-card.types";
import { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { ADDITIONAL_CARDS_DEPENDANT_ROLES } from "~/composables/api/role/constants/role.constants";
import type { RoleName } from "~/composables/api/role/types/role.types";
import { StoreIds } from "~/stores/enums/store.enum";
import { useRolesStore } from "~/stores/role/useRolesStore";
import { useLocalStorage } from "@vueuse/core";
import { LocalStorageKeys } from "~/utils/enums/local-storage.enums";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const rolesStore = useRolesStore();
  const { getRoleWithNameInRoles } = rolesStore;

  const createGameOptionsDtoFromLocalStorage = useLocalStorage(LocalStorageKeys.GAME_OPTIONS, DEFAULT_GAME_OPTIONS, { mergeDefaults: true });

  const defaultCreateGameDto = CreateGameDto.create({
    players: [],
    options: DEFAULT_GAME_OPTIONS,
  });

  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({
    ...defaultCreateGameDto,
    options: createGameOptionsDtoFromLocalStorage.value,
  }));

  const createGameOptionsDto = computed<GameOptions>(() => createGameDto.value.options);

  const doesCreateGameDtoContainPositionDependantRoles = computed<boolean>(() => {
    const positionDependantRoles: RoleName[] = ["rusty-sword-knight", "bear-tamer", "fox"];
    const playersWithPositionDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto(positionDependantRoles);

    return playersWithPositionDependantRoles.length > 0;
  });

  const doesCreateGameDtoContainAdditionalCardsDependantRoles = computed<boolean>(() => {
    const playersWithAdditionalCardsDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto([...ADDITIONAL_CARDS_DEPENDANT_ROLES]);

    return playersWithAdditionalCardsDependantRoles.length > 0;
  });

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function resetCreateGameDto(doesRetrieveLocalStorageValues = true): void {
    if (!doesRetrieveLocalStorageValues) {
      createGameDto.value = CreateGameDto.create(defaultCreateGameDto);
      saveCreateGameOptionsDtoToLocalStorage();

      return;
    }
    createGameDto.value = CreateGameDto.create({
      ...defaultCreateGameDto,
      options: createGameOptionsDtoFromLocalStorage.value,
    });
  }

  function resetCreateGameOptionsDto(): void {
    createGameDto.value.options = GameOptions.create(DEFAULT_GAME_OPTIONS);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function saveCreateGameOptionsDtoToLocalStorage(): void {
    createGameOptionsDtoFromLocalStorage.value = createGameDto.value.options;
  }

  function removeObsoleteAdditionalCardsFromCreateGameDto(): void {
    if (!createGameDto.value.additionalCards) {
      return;
    }
    const rolesInGame = new Set(createGameDto.value.players.map(player => player.role.name));
    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(({ recipient }) => rolesInGame.has(recipient));
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }

  function isRoleInCreateGameDto(roleName: RoleName): boolean {
    return createGameDto.value.players.some(player => player.role.name === roleName);
  }

  function getPlayersWithRoleNameInCreateGameDto(roleName: RoleName): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name === roleName);
  }

  function getPlayersWithAnyRoleNameInCreateGameDto(roleNames: RoleName[]): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name && roleNames.includes(player.role.name));
  }

  function isRoleMinReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && (role.minInGame === undefined || playersWithRoleName.length >= role.minInGame);
  }

  function isRoleMaxReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && playersWithRoleName.length >= role.maxInGame;
  }

  function getRoleLeftCountToReachMinInCreateGameDto(roleName: RoleName): number {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);
    if (role?.minInGame === undefined) {
      return 0;
    }
    const leftCount = role.minInGame - playersWithRoleName.length;

    return leftCount <= -1 ? 0 : leftCount;
  }

  function setAdditionalCardsForRecipientInCreateGameDto(additionalCards: CreateGameAdditionalCardDto[], recipient: GameAdditionalCardRecipientRoleName): void {
    if (!createGameDto.value.additionalCards) {
      createGameDto.value.additionalCards = additionalCards;
    }

    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(card => card.recipient !== recipient);
    createGameDto.value.additionalCards.push(...additionalCards);
  }

  function getAdditionalCardsForRecipientInCreateGameDto(recipient: GameAdditionalCardRecipientRoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.recipient === recipient) ?? [];
  }

  function getAdditionalCardsWithRoleNameInCreateGameDto(roleName: RoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.roleName === roleName) ?? [];
  }
  return {
    createGameOptionsDtoFromLocalStorage,
    createGameDto,
    createGameOptionsDto,
    doesCreateGameDtoContainPositionDependantRoles,
    doesCreateGameDtoContainAdditionalCardsDependantRoles,
    setCreateGameDto,
    resetCreateGameDto,
    resetCreateGameOptionsDto,
    saveCreateGameOptionsDtoToLocalStorage,
    removeObsoleteAdditionalCardsFromCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
    isRoleInCreateGameDto,
    getPlayersWithRoleNameInCreateGameDto,
    getPlayersWithAnyRoleNameInCreateGameDto,
    isRoleMinReachedInCreateGameDto,
    isRoleMaxReachedInCreateGameDto,
    getRoleLeftCountToReachMinInCreateGameDto,
    setAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsWithRoleNameInCreateGameDto,
  };
});

export { useCreateGameDtoStore };