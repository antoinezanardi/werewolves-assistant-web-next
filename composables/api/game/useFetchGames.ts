import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Game } from "~/composables/api/game/types/game.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchGames = {
  createGame: (createGameDto: CreateGameDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  getGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
};

function useFetchGames(): UseFetchGames {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function createGame(createGameDto: CreateGameDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games`, {
        method: "POST",
        body: JSON.stringify(createGameDto),
      });
    } catch (error) {
      return null;
    }
  }

  async function getGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "GET" });
    } catch (error) {
      return null;
    }
  }
  return {
    createGame,
    getGame,
  };
}

export { useFetchGames };