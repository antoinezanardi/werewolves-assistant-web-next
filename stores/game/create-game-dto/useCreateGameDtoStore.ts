import { defineStore } from "pinia";
import { ref } from "vue";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import { StoreIds } from "~/stores/enums/store.enum";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({ players: [] }));

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
  }

  function resetCreateGameDto(): void {
    createGameDto.value = CreateGameDto.create({ players: [] });
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }
  return {
    createGameDto,
    setCreateGameDto,
    resetCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
  };
});

export { useCreateGameDtoStore };