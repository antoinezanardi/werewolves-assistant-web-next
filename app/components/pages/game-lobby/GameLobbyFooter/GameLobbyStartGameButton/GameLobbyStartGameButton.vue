<template>
  <div
    id="game-lobby-start-game-button-container"
    v-p-tooltip.top="containerTooltip"
    class="h-full"
  >
    <PrimeVueButton
      id="game-lobby-start-game-button"
      class="!h-full !px-6 !py-4 start-game-button"
      :disabled="isButtonDisabled"
      :label="$t('components.GameLobbyStartGameButton.startGame')"
      :loading="isLoadingCreateGame"
      raised
      size="large"
      type="button"
      @click.prevent="onClickFromStartGameButton"
    >
      <template #icon>
        <FontAwesomeIcon
          icon="play"
        />
      </template>
    </PrimeVueButton>

    <GameLobbyStartGameConfirmDialog
      id="game-lobby-start-game-confirm-dialog"
      ref="gameLobbyStartGameConfirmDialog"
      @confirm-start-game="onConfirmStartGameFromGameLobbyStartGameConfirmDialog"
      @reject-actor-additional-cards-placed-step="onRejectActorAdditionalCardsPlacedStepFromGameLobbyStartGameConfirmDialog"
      @reject-game-options-changed-step="onRejectGameOptionsChangedStep"
      @reject-players-position-step="onRejectPlayersPositionStepFromGameLobbyStartGameConfirmDialog"
      @reject-thief-additional-cards-placed-step="onRejectThiefAdditionalCardsPlacedStepFromGameLobbyStartGameConfirmDialog"
    />
  </div>
</template>

<script lang="ts" setup>
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import { storeToRefs } from "pinia";

import type { GameLobbyStartGameButtonEmits } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/game-lobby-start-game-button.types";
import type { GameLobbyStartGameConfirmDialogExposed } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/game-lobby-start-game-confirm-dialog.types";
import GameLobbyStartGameConfirmDialog from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialog.vue";
import { useCreateGameDtoValidation } from "~/composables/api/game/useCreateGameDtoValidation";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { usePrimeVueToasts } from "~/composables/prime-vue/usePrimeVueToasts";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const emit = defineEmits<GameLobbyStartGameButtonEmits>();

const gameLobbyStartGameConfirmDialog = ref<GameLobbyStartGameConfirmDialogExposed | null>(null);

const { t } = useI18n();

const { addSuccessToast } = usePrimeVueToasts();

const { createGame } = useFetchGames();

const createGameDtoStore = useCreateGameDtoStore();
const { sanitizeCreateGameDtoForGameCreation } = createGameDtoStore;
const { createGameDto } = storeToRefs(createGameDtoStore);
const { canCreateGame, gameCreationValidationErrors } = useCreateGameDtoValidation(createGameDto);

const isLoadingCreateGame = ref<boolean>(false);

const containerTooltip = computed<string | undefined>(() => gameCreationValidationErrors.value[0]);

const isButtonDisabled = computed<boolean>(() => !canCreateGame.value || isLoadingCreateGame.value);

function onClickFromStartGameButton(): void {
  if (!gameLobbyStartGameConfirmDialog.value) {
    throw createError("Game Lobby Start Game Confirm Dialog is not defined");
  }
  gameLobbyStartGameConfirmDialog.value.open();
}

async function onConfirmStartGameFromGameLobbyStartGameConfirmDialog(): Promise<void> {
  isLoadingCreateGame.value = true;
  sanitizeCreateGameDtoForGameCreation();
  const createdGame = await createGame(createGameDto.value);
  if (createdGame) {
    await navigateTo(`/game/${createdGame._id}`);
    addSuccessToast({ summary: t("components.GameLobbyStartGameButton.gameCreated") });
  }
  isLoadingCreateGame.value = false;
}

function onRejectPlayersPositionStepFromGameLobbyStartGameConfirmDialog(): void {
  emit("rejectPlayersPositionStep");
}

function onRejectThiefAdditionalCardsPlacedStepFromGameLobbyStartGameConfirmDialog(): void {
  emit("rejectThiefAdditionalCardsPlacedStep");
}

function onRejectActorAdditionalCardsPlacedStepFromGameLobbyStartGameConfirmDialog(): void {
  emit("rejectActorAdditionalCardsPlacedStep");
}

function onRejectGameOptionsChangedStep(): void {
  emit("rejectGameOptionsChangedStep");
}
</script>