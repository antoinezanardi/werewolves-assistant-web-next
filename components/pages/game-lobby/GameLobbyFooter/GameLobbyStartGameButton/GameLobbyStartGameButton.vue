<template>
  <div
    id="game-lobby-start-game-button-container"
    v-p-tooltip.top="containerTooltip"
    class="h-full"
  >
    <VuePrimeButton
      class="!h-full start-game-button"
      :disabled="!canCreateGame"
      icon="fa-play fa"
      :label="$t('components.GameLobbyStartGameButton.startGame')"
      :loading="isLoadingCreateGame"
      raised
      size="large"
      type="button"
      @click="handleStartGameButtonClick"
    />
  </div>
</template>

<script lang="ts" setup>
import { storeToRefs } from "pinia";

import { useCreateGameDtoValidation } from "~/composables/api/game/useCreateGameDtoValidation";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { useVuePrimeToasts } from "~/composables/vue-prime/useVuePrimeToasts";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const { t } = useI18n();

const { addSuccessToast } = useVuePrimeToasts();

const { createGame } = useFetchGames();

const createGameDtoStore = useCreateGameDtoStore();
const { createGameDto } = storeToRefs(createGameDtoStore);
const { canCreateGame, gameCreationValidationErrors } = useCreateGameDtoValidation(createGameDto);

const isLoadingCreateGame = ref<boolean>(false);

const containerTooltip = computed<string | undefined>(() => gameCreationValidationErrors.value[0]);

async function handleStartGameButtonClick(): Promise<void> {
  isLoadingCreateGame.value = true;
  const createdGame = await createGame(createGameDto.value);
  if (createdGame) {
    await navigateTo(`/game/${createdGame._id}`);
    addSuccessToast({ summary: t("components.GameLobbyStartGameButton.gameCreated") });
  }
  isLoadingCreateGame.value = false;
}
</script>