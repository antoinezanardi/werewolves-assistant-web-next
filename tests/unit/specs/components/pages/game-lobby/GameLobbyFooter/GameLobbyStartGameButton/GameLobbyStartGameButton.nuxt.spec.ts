import { flushPromises } from "@vue/test-utils";
import type { mount } from "@vue/test-utils";
import type Button from "primevue/button";
import type { Mock } from "vitest";
import type { Ref } from "vue";

import GameLobbyStartGameButton from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameButton.vue";
import * as UseFetchGames from "~/composables/api/game/useFetchGames";
import * as UseVuePrimeToasts from "~/composables/vue-prime/useVuePrimeToasts";
import { RoleNames, RoleSides } from "~/composables/api/role/enums/role.enums";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";
import { createFakeCreateGamePlayerDto } from "~/tests/unit/utils/factories/composables/api/game/dto/create-game/create-game-player/create-game-player.dto.factory";
import { createFakeCreateGameDto } from "~/tests/unit/utils/factories/composables/api/game/dto/create-game/create-game.dto.factory";
import { createFakeGame } from "~/tests/unit/utils/factories/composables/api/game/game.factory";
import { createFakeUseVuePrimeToasts } from "~/tests/unit/utils/factories/composables/vue-prime/useVuePrimeToasts.factory";
import { pTooltipDirectiveBinder } from "~/tests/unit/utils/helpers/directive.helpers";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Game Lobby Start Game Button Component", () => {
  const validCreateGameDto = createFakeCreateGameDto({
    players: [
      createFakeCreateGamePlayerDto({
        name: "Player 1",
        role: { name: RoleNames.VILLAGER },
        side: { current: RoleSides.VILLAGERS },
      }),
      createFakeCreateGamePlayerDto({
        name: "Player 2",
        role: { name: RoleNames.WEREWOLF },
        side: { current: RoleSides.WEREWOLVES },
      }),
      createFakeCreateGamePlayerDto({
        name: "Player 3",
        role: { name: RoleNames.WEREWOLF },
        side: { current: RoleSides.WEREWOLVES },
      }),
      createFakeCreateGamePlayerDto({
        name: "Player 4",
        role: { name: RoleNames.WEREWOLF },
        side: { current: RoleSides.WEREWOLVES },
      }),
    ],
  });

  let wrapper: ReturnType<typeof mount<typeof GameLobbyStartGameButton>>;
  let mocks: {
    composables: {
      useFetchGames: {
        createGame: Mock;
        getGame: Mock;
        cancelGame: Mock;
      };
      useVuePrimeToasts: ReturnType<typeof createFakeUseVuePrimeToasts>;
    }
  };

  beforeEach(async() => {
    mocks = {
      composables: {
        useFetchGames: {
          createGame: vi.fn(),
          getGame: vi.fn(),
          cancelGame: vi.fn(),
        },
        useVuePrimeToasts: createFakeUseVuePrimeToasts(),
      },
    };
    vi.spyOn(UseFetchGames, "useFetchGames").mockReturnValue(mocks.composables.useFetchGames);
    vi.spyOn(UseVuePrimeToasts, "useVuePrimeToasts").mockReturnValue(mocks.composables.useVuePrimeToasts);
    wrapper = await mountSuspendedComponent(GameLobbyStartGameButton);
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Button container", () => {
    it("should assign tooltip when button is disabled.", async() => {
      const tooltip = { value: undefined };
      const directives = { ...pTooltipDirectiveBinder(tooltip, "#game-lobby-start-game-button-container") };
      wrapper = await mountSuspendedComponent(GameLobbyStartGameButton, { global: { directives } });
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto = createFakeCreateGameDto({ players: [] });
      await nextTick();

      expect(tooltip.value).toBe("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached");
    });

    it("should not assign tooltip when button is enabled.", async() => {
      const tooltip = { value: undefined };
      const directives = { ...pTooltipDirectiveBinder(tooltip, "#game-lobby-start-game-button-container") };
      wrapper = await mountSuspendedComponent(GameLobbyStartGameButton, { global: { directives } });
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto = validCreateGameDto;
      await nextTick();

      expect(tooltip.value).toBeUndefined();
    });
  });

  describe("Button", () => {
    it("should be disabled when minimum players are not reached.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto = createFakeCreateGameDto({ players: [] });
      await nextTick();

      const button = wrapper.find(".start-game-button");
      expect(button.attributes("disabled")).toBe("true");
    });

    it("should be enabled when game can be created.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto = validCreateGameDto;
      await nextTick();

      const button = wrapper.find(".start-game-button");
      expect(button.attributes("disabled")).toBe("false");
    });

    describe("Click on button", () => {
      beforeEach(() => {
        const createGameDtoStore = useCreateGameDtoStore();
        createGameDtoStore.createGameDto = createFakeCreateGameDto({
          players: [
            createFakeCreateGamePlayerDto({ name: "Player 1" }),
            createFakeCreateGamePlayerDto({ name: "Player 2" }),
            createFakeCreateGamePlayerDto({ name: "Player 3" }),
            createFakeCreateGamePlayerDto({ name: "Player 4" }),
          ],
        });
      });

      it("should create game when clicked.", async() => {
        const createGameDtoStore = useCreateGameDtoStore();
        const button = wrapper.find(".start-game-button");
        await button.trigger("click");

        expect(mocks.composables.useFetchGames.createGame).toHaveBeenCalledExactlyOnceWith(createGameDtoStore.createGameDto);
      });

      it("should navigate to game with id in params when game is created.", async() => {
        const createdGame = createFakeGame();
        mocks.composables.useFetchGames.createGame.mockResolvedValue(createdGame);
        const button = wrapper.find(".start-game-button");
        await button.trigger("click");
        await flushPromises();

        expect(navigateTo).toHaveBeenCalledExactlyOnceWith(`/game/${createdGame._id}`);
      });

      it("should add success toast when game is created.", async() => {
        mocks.composables.useFetchGames.createGame.mockResolvedValue(createFakeGame());
        const button = wrapper.find(".start-game-button");
        await button.trigger("click");
        await flushPromises();

        expect(mocks.composables.useVuePrimeToasts.addSuccessToast).toHaveBeenCalledExactlyOnceWith({ summary: "components.GameLobbyStartGameButton.gameCreated" });
      });

      it("should not add success toast when game is not created.", async() => {
        mocks.composables.useFetchGames.createGame.mockResolvedValue(null);
        const button = wrapper.find(".start-game-button");
        await button.trigger("click");

        expect(mocks.composables.useVuePrimeToasts.addSuccessToast).not.toHaveBeenCalled();
      });

      it("should be loading when fetch is in progress.", () => {
        const button = wrapper.findComponent<Button>(".start-game-button");
        void button.trigger("click");

        expect((wrapper.vm as unknown as { isLoadingCreateGame: Ref<boolean> }).isLoadingCreateGame.value).toBeTrue();
      });

      it("should not be loading when fetch returned.", async() => {
        const button = wrapper.findComponent<Button>(".start-game-button");
        await button.trigger("click");

        expect(button.attributes("loading")).toBe("false");
      });
    });
  });
});