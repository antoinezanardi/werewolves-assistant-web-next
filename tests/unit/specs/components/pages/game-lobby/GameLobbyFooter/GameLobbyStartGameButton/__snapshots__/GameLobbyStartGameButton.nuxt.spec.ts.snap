// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Start Game Button Component > should match snapshot when rendered. 1`] = `
"<div id="game-lobby-start-game-button-container" class="h-full" data-pd-tooltip="true">
  <button-stub id="game-lobby-start-game-button" class="!h-full !px-6 !py-4 start-game-button" disabled="true" label="Start game" loading="false" raised="" size="large" type="button"></button-stub>
  <div id="game-lobby-start-game-confirm-dialog"></div>
</div>"
`;
